# A simple example of a joint distribution

In [89]:
import numpy as np

In [90]:
def rr(x):
    return np.round(x, 3)

#### A joint distribution $f_{XY}(x,y)$ 

In [134]:
# Assume the support of Y is {1,2} and the supporrt of X is {1,2,3} (careful that Python indexing starts at 0 below)
f_XY = np.array([[0.1, 0.3, 0.2],
                 [0.2, 0.1, 0.1]])
print(f_XY)

[[ 0.1  0.3  0.2]
 [ 0.2  0.1  0.1]]


#### The marginal distribution $f_X(x) = \sum_y f_{XY}(x,y)$

In [164]:
f_X = rr(np.sum(f_XY, axis = 0))
f_X

array([ 0.3,  0.4,  0.3])

#### The marginal distribution $f_Y(y) = \sum_x f_{XY}(x,y)$

In [93]:
f_Y = rr(np.sum(f_XY, axis = 1))
f_Y

array([ 0.6,  0.4])

#### The conditional distribution $f_{X|Y}(x|y) = \frac{ f_{XY}(x,y)}{f_Y(y)}$

In [94]:
f_XgivenY = rr(f_XY / f_Y[:,np.newaxis])
f_XgivenY

array([[0.167,  0.5, 0.333],
       [ 0.5, 0.25, 0.25]])

##### The case conditional on the first Y

In [95]:
f_XgivenY[0,:]

array([0.167,  0.5, 0.333])

##### The case conditional on the second Y

In [96]:
f_XgivenY[1:]

array([[ 0.5, 0.25, 0.25]])

#### The conditional distribution $f_{Y|X}(y|x) = \frac{ f_{XY}(x,y)}{f_X(x)}$

In [97]:
f_YgivenX = rr(f_XY / f_X[np.newaxis, :])
f_YgivenX

array([[0.333, 0.75, 0.667],
       [0.667, 0.25, 0.333]])

Here each column of the above is a conditional distribution of $Y$ given $X$

#### Moments

$$
{\mathbb E}[X] = \sum_x x f_X(x),
\qquad
{\mathbb E}[Y] = \sum_y y f_Y(y)
$$

##### Mean

In [98]:
E_X = float(np.sum(f_X * np.arange(1,4)))
E_X

2.0

In [99]:
E_Y = float(np.sum(f_Y * np.arange(1,3)))
E_Y

1.4

The mean vector:
$$
\mu = 
\left[
\begin{matrix}
{\mathbb E}[X] \\
{\mathbb E}[Y]
\end{matrix}
\right]
$$

In [100]:
mu = np.array([E_X, E_Y])
mu

array([ 2.0,  1.4])

##### Variance

In [101]:
E_X2 = float(np.sum(f_X * np.arange(1,4)**2)) # second moment
Var_X = E_X2 - E_X**2
round(Var_X, 3)

0.6

In [102]:
E_Y2 = float(np.sum(f_Y * np.arange(1,3)**2)) # second moment
Var_Y = E_Y2 - E_Y**2
round(Var_Y,3)

0.24

##### Covariance

$$
\text{Cov}(X,Y) ={\mathbb E}[(X- {\mathbb E}[X])(Y- {\mathbb E}[Y])] = {\mathbb E}[XY] -{\mathbb E}[X] {\mathbb E}[Y]
$$

$$
{\mathbb E}[XY] = \sum_{x,y} x y f_{XY}(x,y)
$$

In [142]:
np.outer([1,2],[1,2,3])

array([[   1,    2,    3],
       [   2,    4,    6]])

In [143]:
E_XY = np.sum(np.outer([1,2],[1,2,3]) * f_XY)
float(E_XY)

2.7

In [144]:
cov_XY = E_XY - E_X*E_Y
float(round(cov_XY, 3))

-0.1

##### Correlation Coefficient

$$
\rho(X,Y) = \frac{\text{Cov}(X,Y)}{\sqrt{\text{Var}(X)\text{Var}(Y)}}
$$

In [146]:
rho = cov_XY/np.sqrt(Var_X*Var_Y)
float(np.round(rho,3))

-0.264

### Conditional expectation

In [154]:
f_XgivenY

array([[0.167,  0.5, 0.333],
       [ 0.5, 0.25, 0.25]])

In [155]:
E_XgivenYis1 = np.sum(f_XgivenY[0,:] * np.arange(1,4))
float(rr(E_XgivenYis1))

2.166

In [156]:
E_XgivenYis2 = np.sum(f_XgivenY[1,:] * np.arange(1,4))
float(E_XgivenYis2)

1.75

In [157]:
f_Y

array([ 0.6,  0.4])

The random variable ${\mathbb E}[X ~|~Y]$ has a distribution with probability mass function

$$
f(x) = \begin{cases}
0.6 & x = 2.166, \\
0.4 & x = 1.75, \\
0 & \text{otherwise}.
\end{cases}
$$

##### The Tower Property

$$
{\mathbb E}[X] = {\mathbb E} \Big[ {\mathbb E}[X ~|~ Y] \Big] 
$$

In [166]:
float(rr(E_XgivenYis1 * f_Y[0] + E_XgivenYis2 * f_Y[1]))

2.0

So we retrieve the expecation of $X$ that we had before.

#### Independence

In [173]:
f_XY_ind = rr(np.outer(f_Y, f_X))
f_XY_ind

array([[0.18, 0.24, 0.18],
       [0.12, 0.16, 0.12]])

In [176]:
# f_XY_ind NOt the same as f_XY
f_XY

array([[ 0.1,  0.3,  0.2],
       [ 0.2,  0.1,  0.1]])

In [177]:
E_XY_ind = np.sum(np.outer([1,2],[1,2,3]) * f_XY_ind)
float(E_XY_ind)

2.8

In case of independence the covariance must be 0.

In [179]:
cov_XY_ind = E_XY_ind - E_X*E_Y
float(cov_XY_ind)

0.0